carla/recourse_methods/catalog/revise/model.py

```diff
@@ -0,0 +1,206 @@
+from typing import Dict
+
+import numpy as np
+import pandas as pd
+import torch
+from torch import nn
+
+from carla.data.api import Data
+from carla.models.api import MLModel
+from carla.recourse_methods.api import RecourseMethod
+from carla.recourse_methods.autoencoder import (
+    Dataloader,
+    VariationalAutoencoder,
+    train_variational_autoencoder,
+)
+from carla.recourse_methods.processing.counterfactuals import (
+    check_counterfactuals,
+    reconstruct_encoding_constraints,
+)
+
+
+class Revise(RecourseMethod):
+    def __init__(self, mlmodel: MLModel, data: Data, hyperparams: Dict) -> None:
+        """
+        Initialisation of the REVISE recourse method.
+
+        Restrictions
+        ------------
+        - Works currently only on Pytorch models
+        - Only binary categorical features with and without one-hot-encoding
+
+        Parameters
+        ----------
+        mlmodel: Black-box-model we want to explore
+        data: Dataset to perform on
+        hyperparams: Parameter for Revise method, with following possibilites
+            {
+                "data_name": str  name of the dataset,
+                "lambda": float default: 0.5    Decides how similar the counterfactual is to the factual,
+                "optimizer": str defaul: "adam" Optimizer for generation of counterfactuals,
+                            only adam and rmsprop possible
+                "lr": float default: 0.1    learning rate for Revise,
+                "max_iter": int default 1000, number of iterations for Revise optimization,
+                "target_class": List default: [0, 1]  List of one-hot-encoded target class,
+                "binary_cat_features": bool default: True If true, the encoding of x is done by drop_if_binary
+                "vae_params": Dict with parameter for VAE,
+                    {
+                        "d": 8,  # latent space
+                        "D": test_factual.shape[1],  # input size
+                        "H1": 512,
+                        "H2": 256,
+                        "train": False,
+                        "lambda_reg": 1e-6,
+                        "epochs": 5,
+                        "lr": 1e-3,
+                        "batch_size": 32,
+                    }
+            }
+        """
+        super().__init__(mlmodel)
+        self.params = hyperparams
+
+        self._target_column = data.target
+        self._lambda = self.params["lambda"] if "lambda" in hyperparams.keys() else 0.5
+        self._optimizer = (
+            self.params["optimizer"] if "optimizer" in hyperparams.keys() else "adam"
+        )
+        self._lr = self.params["lr"] if "lr" in hyperparams.keys() else 0.1
```

Comment:
At some point we should think of a smarter way to load defaults, seems like a re-occuring pattern that we have to have 10 if else statements in the constructor of each recourse method

Reply:
That's right, I'll have a look into it while I'm refactoring